### ISSUE TITLE
embedding layer with LSTM autoencoder

### ISSUE THREAD

**Issue by mg64ve** [rstudio/keras#583]
Hello, 

I have found the following answer but I want to make sure it really does not have any other option:

https://stackoverflow.com/questions/44731059/keras-lstm-autoencoder-with-embedding-layer

The following use of embedding layer with timeseries works:

```
num_features = 4
nb_classes =3
num_steps =10


X <- matrix(0,nrow=128,ncol=num_features)
y <- matrix(0,nrow=128,ncol=nb_classes)


model <<- keras_model_sequential()
model %>%
layer_embedding(num_features, num_steps) %>%
layer_lstm(units = num_features, return_sequences=FALSE) %>% 
layer_dense(units=nb_classes, activation='linear')

model %>% compile(loss='categorical_crossentropy', optimizer='rmsprop')

summary(model)

model %>% fit(X, y, batch_size = 1, epochs = 1, verbose = 1)
```
Similar concept with LSTM autoencoder does not work:

```
num_features = 4
num_steps = 10


X <- matrix(0,nrow=128,ncol=num_features)

model <<- keras_model_sequential()
model %>%
layer_embedding(num_features, num_steps) %>%
layer_lstm(units = num_features, return_sequences = FALSE) %>% 
layer_repeat_vector(num_steps) %>% 
layer_lstm(units = num_features, return_sequences = TRUE) %>%
time_distributed(layer_dense(units = num_features))


model %>% compile(optimizer = optimizer_rmsprop(lr = 0.0001, decay=0.000), loss='mse')	

summary(model)

model %>% fit(X, X, batch_size = 1, epochs = 1, verbose = 1)
```

The issue is that AE uses X as input and output and X is not embedded.
Is there any solution for this issue?

Thanks.
Cheers.

**Comment by skeydan** [comment]
I'm not aware of any general solutions to this other than workarounds in the line of what's suggested in the SO issue.

However, I'd also ask if it's worth the effort for your specific application... if it's an NLP task, I guess it makes sense to want to use embeddings. So in that case, perhaps you could split up the task into 2 parts, first learn the embeddings separately and then, code an autoencoder that takes these as inputs and outputs both.

On the other hand, dependent on the data you're working with, embeddings may not add _too_ much in terms of accuracy, so perhaps I'd try without first and see how far that gets me (it also seems like LSTM autoencoder does not necessarily achieve everyone's goal, see e.g. https://github.com/keras-team/keras/issues/10753, so I'd check on a low-oeverhead way first & see if it does what you want).

**Comment by mg64ve** [comment]
@skeydan  thanks.

> So in that case, perhaps you could split up the task into 2 parts, first learn the embeddings separately 

This is not clear to me. Could you please give me an example?

**Comment by skeydan** [comment]
I meant training e.g. an encoder-decoder model (like here: https://blogs.rstudio.com/tensorflow/posts/2018-07-30-attention-layer/ - just an example), and then extracting the embeddings and treating them as X as well as Y.

But that's a purely technical suggestion; I don't know how useful an LSTM autoencoder per se would be (will of course depend on what you want to do - just one sees not many applications of it on the net).

**Comment by mckeown12** [comment]
Hi All, I have a good use case for an LSTM autoencoder (namely some IOT timeseries data that comes in the form of irregular length sequences) and seem to be having a [similar problem](https://stats.stackexchange.com/questions/375821/lstm-autoencoder-keras-variable-batch-size)-- I can encode a variable length sequence as a fixed length vector, but then the `layer_repeat_vector(num_steps)` trick used above doesn't work for the decoder.  Any ideas?

**Comment by mg64ve** [comment]
Hi, what error do you get?
By the way this ticket was open for another kind of problem.

**Comment by mckeown12** [comment]
Sorry, I see the original ticket was about the embedding layer, so this is likely the wrong place for me to be posting. I'm not getting an error, this is just a conceptual problem-- if I repeat the latent vector a constant `num_steps`  times, I won't get output of the right shape to compute loss with or do backprop since my input sequences are variable length.  I know that in pytorch uses packing and padding of sequences to deal with this, but I don't see an obvious way to proceed with a static computational graph.

**Comment by mg64ve** [comment]
maybe @skeydan can give you a more detailed answer
I think you need to use a masking layer
But I am not sure, I have never used.

**Comment by skeydan** [comment]
I think you could use probably use padding as in NLP.
There you also have different-length sentences (words, ...) normally and you pad (or truncate) them to a common length (see e.g. `pad_sequences`).
